src/printf.cpp

```diff
@@ -0,0 +1,249 @@
+
+    // Decompose the remaining format string into individual strings with
+    // one format specifier each, handle each one individually
+    size_t arg_idx = 0;
+    while (next_part < format_string.size() - 1) {
+        // Get the part of the format string before the next format specifier
+        size_t part_start = next_part;
+        size_t part_end = format_string.find_first_of('%', part_start + 1);
+        auto part_fmt = format_string.substr(part_start, part_end - part_start);
+
+        // Handle special cases
+        if (part_end == part_start + 1) {
+            printf_out << "%";
+            next_part = part_end + 1;
+            continue;
+        } else if (part_end == std::string::npos &&
+                   arg_idx >= descs.at(printf_id).arg_sizes.size()) {
+            // If there are no remaining arguments, the rest of the format
+            // should be printed verbatim
+            printf_out << part_fmt;
+            break;
+        }
+
+        // The size of the argument that this format part will consume
+        auto& size = descs.at(printf_id).arg_sizes[arg_idx];
+
+        // Check to see if we have a vector format specifier
+        int vec_len = 0;
+        int el_size = 0;
+        std::string remaining_str;
+        part_fmt = get_vector_fmt(part_fmt, vec_len, el_size, remaining_str);
+
+        // Scalar argument
+        if (vec_len < 2) {
+            // Special case for %s
+            if (get_fmt_conversion(part_fmt) == 's') {
+                uint32_t string_id = read_buff<uint32_t>(data);
+                printf_out << print_part(
+                    part_fmt, descs.at(string_id).format_string.c_str(), size);
+            } else {
+                printf_out << print_part(part_fmt, data, size);
+            }
+            data += size;
+        } else {
+            // Vector argument
+            auto* data_start = data;
+            for (int i = 0; i < vec_len - 1; i++) {
+                printf_out << print_part(part_fmt, data, size / vec_len) << ",";
+                data += el_size;
+            }
+            printf_out << print_part(part_fmt, data, size / vec_len)
+                       << remaining_str;
+            data = data_start + size;
+        }
+
+        // Move to the next format part and prepare to handle the next arg
+        next_part = part_end;
+        arg_idx++;
+    }
+
+    printf("%s", printf_out.str().c_str());
+}
+
+void cvk_printf(cvk_mem* printf_buffer,
+                const printf_descriptor_map_t& descriptors) {
+    if (!printf_buffer->map()) {
+        cvk_error("Could not map printf buffer");
+        return;
+    }
+    char* data = static_cast<char*>(printf_buffer->host_va());
+    auto buffer_size = printf_buffer->size();
+    auto bytes_written = read_inc_buff<uint32_t>(data) * 4;
+    auto* data_start = data;
+
+    while (static_cast<size_t>(data - data_start) < bytes_written &&
+           static_cast<size_t>(data - data_start) < buffer_size) {
```

Comment:
Should this be `buffer_size - 4`?

Reply:
I agree. The 4bytes being for the first element being the buffer offset as I understand it.